lib/models/per-bundle-addon-cache/addon-proxy.js

```diff
@@ -0,0 +1,89 @@
+'use strict';
+
+/**
+ * Returns a proxy to a target with specific handling for the
+ * `parent` property, as well has to handle the `app` property;
+ * that is, the proxy should maintain correct local state in
+ * closure scope for the `app` property if it happens to be set
+ * by `ember-cli`. Other than `parent` & `app`, this function also
+ * proxies _almost_ everything to `target.instance` with a few
+ * exceptions: we trap & return `[]` for `addons`, and we don't return
+ * the original `included` (it's already called on the "real" addon
+ * by `ember-cli`).
+ *
+ * Note: the target is NOT the singleton instance of the addon. Rather,
+ * it is a CacheEntry from SingletonAddonCache.
+ *
+ * @method getAddonProxy
+ * @param targetCacheEntry the PerBundleAddonCache CacheEntry we are to proxy
+ * @param parent the parent object of the proxy being created (the same as
+ * the 'parent' property of a normal addon instance)
+ * @return Proxy
+ */
+function getAddonProxy(targetCacheEntry, parent) {
+  let _app;
+  let addonProxy = new Proxy(targetCacheEntry, {
+    get(targetCacheEntry, property) {
+      if (property === 'parent') {
+        return parent;
+      }
+
+      if (property === 'app') {
+        return _app;
+      }
+
+      if (property === 'isProxy') {
+        return true;
+      }
+
+      // trap addons & don't return the real addons array; this is done for primarily
+      // two reasons: (1) `ember-engines` dedupe logic is stateful, so this will just
+      // completely break, and (2) with all dependencies declared explicitly, the build
+      // is _extremely_ slow (eg, it takes over 40 minutes!)
+      if (property === 'addons') {
+        return [];
+      }
+
+      // allow access to `instance`
+      if (property === 'instance') {
+        return targetCacheEntry.instance;
+      }
+
+      // `included` will be called on the "real" addon, so there's no need for it to be
+      // called again; instead we return a no-op implementation here
+      if (property === 'included') {
+        return () => undefined;
+      }
+
+      if (targetCacheEntry.isBuilt) {
+        if (property !== 'constructor' && typeof targetCacheEntry.instance[property] === 'function') {
+          // we don't want the proxy to be the context of these methods, so wrap
+          // the original methods to maintain the correct context
+          return function _originalAddonPropMethodWrapper() {
+            return targetCacheEntry.instance[property](...arguments);
+          };
+        }
```

Comment:
Chatted with @brendenpalmer and @davecombs, the reason for this is somewhat subtle and I'd like to make sure that the comment explains it. Specifically, if we fall through to the `Reflect.get` just below the `this` context of the function when invoked is **the proxy** not the original instance (so its local state is incorrect).

Annnnyyyyyywwwwwaaaaayyyyssssss, lets update the comment to make that clear.